### ISSUE TITLE
Commit leaves inverted index, and Reset jumps back too far

### ISSUE THREAD

**Issue by AArnott** [libgit2/libgit2sharp#1412]
Nerdbank.GitVersioning has tests that pass when run with libgit2sharp 0.22.0, but one test ([GetVersionHeight_VersionJsonHasParsingErrorsInHistory](https://github.com/AArnott/Nerdbank.GitVersioning/blob/master/src/NerdBank.GitVersioning.Tests/GitExtensionsTests.cs#L103)) fails when run against 0.23.1. 

Specifically, it appears that when I call

```csharp
this.Repo.Index.Add(relativeFilePath);
this.Repo.Commit("some text");
```

This leaves the repo in an odd state. The commit is made, but then the inverse of that commit is staged to the index.

Later, when I call Reset to move the head and index back one step but without touching the work tree, 

```csharp
this.Repo.Reset(ResetMode.Mixed, this.Repo.Head.Commits.Skip(1).First());
```

This jumps HEAD all the way back to the start of the repo (2 steps instead of just 1).

**Comment by ethomson** [comment]
That's pretty strange...  there really haven't been that many changes between 0.22 and 0.23.  And as far as libgit2 goes, I can't think of any changes that would change the very basic add + commit scenarios.  And having the index get reverted to `HEAD~1` really doesn't make any sense.

Do you have any other unit tests that do an add and commit that are failing in a similar way?

**Comment by AArnott** [comment]
This is the only one that seems impacted. I suspect it's because it's the only test that uses the reset method. I plan to add a unit test to this repository in an attempt to Repro it.

**Comment by AArnott** [comment]
I have a repro in a simple unit test within this repo. I've pushed the commit that adds it on v0.22 where the test passes. If you merge in v0.23.1 the test fails, as advertised.

I used git bisect with the script below to determine that the commit that introduces the regression is 82d621cda3ea97281873a8543d00ee91577438db:

```
pushd "%~dp0"
git cherry-pick fix1412
git reset --mixed HEAD~1
C:\tools\NuGet.exe restore
msbuild /nologo /m /fl /v:minimal
.\packages\xunit.runner.console.2.0.0\tools\xunit.console.x86.exe .\LibGit2Sharp.Tests\bin\Debug\LibGit2Sharp.Tests.dll -method "LibGit2Sharp.Tests.CommitFixture.CommitLeavesIndexIntact"
git reset --hard
popd
```

**Comment by AArnott** [comment]
BTW, if as you look at the unit test you can think of a workaround I can use in the meantime, I'd be very grateful. I'm trying to adapt Nerdbank.GitVersioning to the portable version of libgit2sharp, but that requires that I take the update that is buggy, which actually won't matter at runtime but for tests it messes me up.

**Comment by ethomson** [comment]
So: `repo.Index.Add(...)` will add a file to the index but it doesn't _write_ the index.  `repo.Commands.Stage` (or just `repo.Stage` in 0.22, deprecated in 0.23) will `Add` and `Write` the index.

So that change isn't actually being committed.

To be completely honest, I'm not entirely sure when this changed.  I wouldn't have thought the libgit2 update would have changed things, but I'll take a closer look.

But as for workarounds, use `Stage` (or `Add` then `Write`) and that should be solid.

**Comment by ethomson** [comment]
Ah, it turns out that the `Index` methods used to update the index after operating, which meant that there was no way to operate on the index in-memory.  So we split things up into `Add` (which is a more "plumbing" method) and `Stage` (which is a more "porcelain" command and does the save).

Looking back on #1291, it looks like we may have done a poor job of documenting this.  Apologies for that.

**Comment by AArnott** [comment]
Thanks for the review, @ethomson. I tried swapping out all the Index commands for the `Commands.Stage` method but the bug still repros.

**Comment by ethomson** [comment]
That's odd, it does not for me.  I added your test to 0.23.1 and it does fail.  But after changing things to `Stage`, it passes.

When you say that the bug repros, are you saying that your LibGit2Sharp test fails, or that it fails within `GitVersioning`?  (Or both.)

**Comment by AArnott** [comment]
> So that change isn't actually being committed.

Yes it is. I can see in git.exe history that the commit is valid. It's only the index after the commit that's messed up.

**Comment by AArnott** [comment]
My LibGit2Sharp test certainly fails. I haven't tried changing everything in NB.GV yet.

Here is my commit that fixes the test to use Stage: 734fc4f9592677e383ec1131aeaf37e2f67865ee. It's the tip of the fix1412 branch in my fork. Can you see what I did differently than you did?

**Comment by AArnott** [comment]
BTW, I really appreciate how easy it is to add a unit test to this library. That's the way to do it. :)

**Comment by carlosmn** [comment]
>>So that change isn't actually being committed.

>Yes it is. I can see in git.exe history that the commit is valid. It's only the index after the commit that's messed up.

This is likely due to some sloppy object lifetime management. What the correct thing to do is a fight for a different day, but what is most likely happening here is that we're re-using the same `git_index` object attached to the repository. By using the `Repository.Index` getter, the `Index.Add()` acts on that instance. Then `Repository.Commit()` takes in that same object and writes it out to a tree, making the commit "valid", or at least the expected value.

Since this index state is never persisted to disk, `git diff` and friends show the old index. You would see the same thing if you re-read the index file yourself. There is no "reversed" index because it doesn't store changes, but represents a state. The index is a commit behind, so it looks like you undid your changes in the index (since it's the same state that you'd get if you did that).

The reset methods (beyond soft) do write out the given commit's tree to the index file, but we don't refresh automagically most of the time, so that might explain why you see some difference.

**Comment by AArnott** [comment]
Thanks, @carlosmn. I'll disregard the state of the index then. But when I leave git.exe out of it and just run the test, it ultimately fails because the reset-mixed command does something different still. Do we have any insight into that at this point?

**Comment by carlosmn** [comment]
My guess would be that `Repository.Commits` underspecifies the order it expects to see commits in. Either way, `HEAD~1` is _not_ the same as `Repository.Commits.Skip(1)` except for some pretty specific circumstances around the ordering of the output. If you want to specify "the parent of the current HEAD" then you should either do it by hand via a lookup and access to its first parent, or by doing the equivalent of `rev-list`.

Can you check whether the output of `this.Repo.Commits` stays the same for the repository you're testing with?

**Comment by ethomson** [comment]
@AArnott This test passes for me in `master`:

```
        [Fact]
        public void CommitLeavesIndexIntact()
        {
            string repoPath = InitNewRepository();

            const string fileName = "a";
            string filePath = Path.Combine(Path.GetDirectoryName(repoPath.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)), "a");
            using (var repo = new Repository(repoPath))
            {
                File.WriteAllText(filePath, "first");
                repo.Stage(fileName);
                var result = repo.Commit("first commit", Constants.Signature, Constants.Signature);
                repo.Reset(ResetMode.Hard, result);

                File.WriteAllText(filePath, "second");
                repo.Stage(fileName);
                result = repo.Commit("second commit", Constants.Signature, Constants.Signature);
                repo.Reset(ResetMode.Hard, result);

                File.WriteAllText(filePath, "third");
                repo.Stage(fileName);
                result = repo.Commit("second commit", Constants.Signature, Constants.Signature);
                repo.Reset(ResetMode.Hard, result);

                Console.WriteLine("Reset mixed to: {0}", repo.Head.Commits.Skip(1).First().Id.Sha);
                repo.Reset(ResetMode.Mixed, repo.Head.Commits.Skip(1).First());
                Assert.Equal(2, repo.Head.Commits.Count());
            }
        }
```

**Comment by AArnott** [comment]
When I merge master into my fix1412 branch, I still get the same failure: Expected 2, Actual 1.

**Comment by AArnott** [comment]
@carlosmn said:

> My guess would be that Repository.Commits underspecifies the order it expects to see commits in.

That appears to be the case. When I examine the order of `repo.Head.Commits` I get "third first second". So it seems that the Commits property used to return commits in parent order (or perhaps tended to by chance) and with newer versions it is more mixed up. Did I make an assumption that was never a guarantee?

When I change my test so that the repo.Reset line is:

```csharp
repo.Reset(ResetMode.Mixed, repo.Head.Tip.Parents.Single());
```

The test passes on v0.23.1

**Comment by carlosmn** [comment]
>So it seems that the Commits property used to return commits in parent order (or perhaps tended to by chance) and with newer versions it is more mixed up. Did I make an assumption that was never a guarantee?

The short answer is "yes".

There is no total order of commits in the general case, and there's no indication of what sorting options a getter like `Repo.Head.Commits` applies without looking at the code (and even looking at it, it doesn't seem to provide any) so unfortunately it does look like it would have to fall under "undocumented" and it just happened to return what you expected.

In the revision walker revamp we did for libgit2 v0.25 we got closer to git and that means that commits get loaded into LIFO  lists a couple of times and then time-sorted. If you do choose the equivalent of `--topo-order` can get something like what you seem to be expecting, but even then there's no single correct way. git's own documentation for `rev-list`'s `--topo-order` lists two possible outputs. If you don't specify a sorting, then all bets are off and we'll just return the commits in *some* order which is more efficient if you don't actually care about the order.

Unfortunately libgit2sharp has its share of assumptions in this area and others. It potentially makes sense to make this particular call time-sorted which is `rev-list`'s default, but even then you wouldn't be able to rely on that to find `HEAD~1`.
